### PR TITLE
fix(buf): prevent EAGAIN hang on OpenBSD non-blocking sockets

### PR DESCRIPTION
**Pull request by DavidOsipov**
## Description
Fixes silent connection hangs on OpenBSD caused by `readv(2)` returning `EAGAIN` on non-blocking sockets for payloads >16KB.

**Root Cause:**
On OpenBSD 7.5+, invoking `SYS_READV` via `syscall.Syscall` with multi-buffer iovecs on non-blocking sockets frequently returns `EAGAIN` even when the OS buffer contains data. Because Xray treats `EAGAIN` as a standard "retry later" signal, this results in an infinite wait (hang) on heavy pages. Additionally, `unix.Readv` is missing for OpenBSD in `golang.org/x/sys v0.41.0`. 

The previous workaround bypassed the issue by falling back to a plain `SingleReader` on OpenBSD, which sacrificed all multi-buffer throughput benefits.

**The Fix (`openbsdSeqReader`):**
This PR replaces the fallback with a platform-specific `multiReader` implementation that relies on a sequential `syscall.Read` loop per buffer. 
* **Preserves adaptive allocation:** Maintains the exact 1→8 buffers × 8 KB strategy used by Linux/macOS.
* **Zero CPU Spin:** Explicitly distinguishes between `EAGAIN` (`return -1`) and fatal errors like `ECONNRESET` (`return 0`), preventing the 100% CPU spin loop in `rawConn.Read`.
* **Short-read optimization:** Breaks the loop early when `n < int(Size)`, minimizing syscall overhead.
* **No CGO / No unsafe:** Uses safe `syscall.Read` which wraps OpenBSD's `libc` internally in modern Go.

## Testing Strategy
* **Unit & Edge Cases:** Covered `EAGAIN` mid-loop, `EOF` mid-loop, short-read breaks, and exact boundary size matrices (8191, 8192, 16384 bytes, etc.).
* **Integration:** Added fragmented stream tests (1ms delays) and 100% CPU spin regression checks using real TCP sockets (`net.Listen`/`net.Dial`, avoiding `net.Pipe` fd panics).
* **Coverage:** Brought test coverage for modified `readv_*.go` files to 100%.
* **Real-World Validation:** Cross-compiled (`GOOS=openbsd GOARCH=arm64 CGO_ENABLED=0`) and verified in production as a routing daemon on OpenBSD 7.8 (NanoPi Neo3).

## Checklist
- [x] I have read the [contributing guidelines](https://github.com/XTLS/Xray-core/blob/main/CONTRIBUTING.md) (if applicable).
- [x] Code compiles correctly across all target platforms (verified `!wasm`, `illumos`, `posix` tags).
- [x] `go test ./common/buf/...` passes on host OS with no regressions.
- [x] Commits are formatted logically and messages are descriptive.

## Detailed documentation for historical purposes:
[HANDOFF.md](https://github.com/user-attachments/files/25846841/HANDOFF.md)

---

## 简介 (Description)
修复了在 OpenBSD 上由于非阻塞套接字在处理大于 16KB 的负载 (payload) 时 `readv(2)` 返回 `EAGAIN` 而导致的连接静默挂起问题。

**根本原因 (Root Cause):**
在 OpenBSD 7.5+ 上，如果在非阻塞套接字上通过 `syscall.Syscall` 调用 `SYS_READV` 并使用多缓冲区 iovecs，即使操作系统缓冲区中仍有数据，也经常会返回 `EAGAIN`。由于 Xray 将 `EAGAIN` 视为标准的“稍后重试”信号，这会导致在加载大型网页时发生无限等待（挂起）。此外，`golang.org/x/sys v0.41.0` 缺少针对 OpenBSD 的 `unix.Readv` 实现。

之前的临时解决方案是在 OpenBSD 上回退到普通的 `SingleReader` 来绕过该问题，但这牺牲了多缓冲区带来的所有吞吐量优势。

**修复方案 (`openbsdSeqReader`):**
本 PR 使用特定平台的 `multiReader` 实现替换了该回退方案，该实现依赖于针对每个缓冲区顺序执行的 `syscall.Read` 循环。
* **保留自适应分配 (Preserves adaptive allocation):** 维持与 Linux/macOS 完全相同的 1→8 缓冲区 × 8 KB 策略。
* **零 CPU 空转 (Zero CPU Spin):** 明确区分 `EAGAIN`（`返回 -1`）和诸如 `ECONNRESET` 的致命错误（`返回 0`），防止 `rawConn.Read` 中出现 100% CPU 空转死循环。
* **短读取优化 (Short-read optimization):** 当 `n < int(Size)` 时提前中断循环，最小化系统调用开销。
* **无 CGO / 无 unsafe 代码 (No CGO / No unsafe):** 使用安全的 `syscall.Read`，在现代 Go 中它内部封装了 OpenBSD 的 `libc`。

## 测试策略 (Testing Strategy)
* **单元测试与边界情况 (Unit & Edge Cases):** 覆盖了循环中途的 `EAGAIN`、循环中途的 `EOF`、短读取中断，以及精确的边界大小矩阵（8191、8192、16384 字节等）。
* **集成测试 (Integration):** 使用真实的 TCP 套接字（`net.Listen`/`net.Dial`，避免了 `net.Pipe` 的 fd panic 问题）添加了碎片化流测试（1毫秒延迟）和 100% CPU 空转的回归检查。
* **代码覆盖率 (Coverage):** 将修改后的 `readv_*.go` 文件的测试覆盖率提升至 100%。
* **实际环境验证 (Real-World Validation):** 完成交叉编译（`GOOS=openbsd GOARCH=arm64 CGO_ENABLED=0`）并在 OpenBSD 7.8 (NanoPi Neo3) 上作为路由守护进程在生产环境中验证通过。

## 检查清单 (Checklist)
- [x] 我已阅读 [贡献指南](https://github.com/XTLS/Xray-core/blob/main/CONTRIBUTING.md)（如适用）。
- [x] 代码在所有目标平台上均能正确编译（已验证 `!wasm`、`illumos`、`posix` 标签）。
- [x] `go test ./common/buf/...` 在主机操作系统上通过，无任何回归问题。
- [x] 提交 (Commits) 格式符合逻辑，且提交信息描述清晰。

### CODE REVIEW

File: common/buf/readv_openbsd.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+			return 0
+		}
+		if n == 0 {
+			if total > 0 {
+				// Deliver partial data first; EOF will be observed on the next call.
+				return total
+			}
+			// Clean EOF when no bytes were read in this round.
+			return 0
+		}
```
[the remark this answers is not in the document]
Comment (DavidOsipov) [reply common/buf/readv_openbsd.go]: Fixed

**Comment by Fangliding** [comment]
我觉得直接在openbsd上直接禁用readv应该比接受一个由纯ai编写我们无法验证的reader更好(这个问题那么久只有少量反馈无效反馈侧面印证了这个系统的用户少得可怜(

**Comment by DavidOsipov** [comment]
Hi @Fangliding,

I understand decision. From a maintainer's perspective, this code might be a risk.

Just a quick clarification on the "pure AI" part: while I did use AI to write code, help format this PR, generate the test boilerplate, and translate my text, the patch itself was born out of a real physical issue. It was successfully cross-compiled and verified in production on my physical NanoPi Neo3 running OpenBSD 7.8, completely fixing the `EAGAIN` hangs caused by the old fallback. 

I will maintain this patch in my local fork for my own server.

**Comment by Fangliding** [comment]
我没有质疑它不能work（因为我也没有设备） 只是根据你的描述我实在不能确定你是否清楚问题所在 而且代码也有问题

**Comment by DavidOsipov** [comment]
I will be completely transparent: I am not a systems programmer by trade; I am a Product Manager. My core expertise is not writing Go syntax, but rather system analysis, isolating variables, and tracking down root causes. 

Because I cannot write low-level Go code myself, I utilized AI as my "Developer", but I guided it with extremely strict architectural constraints based entirely on my own physical investigation.

Here is exactly how the process worked (I am attaching two documents that log this in detail):

1. **Diagnostics (see [xray-openbsd-bug.md](https://github.com/user-attachments/files/25847974/xray-openbsd-bug.md)):** I spent 14+ hours systematically ruling out false trails on my actual hardware (PF firewall, Nginx, MTU, IPv6, `login.conf` limits). I isolated Xray on a separate port and proved the issue was strictly at the TCP stream reading level. I localized the exact bug: on OpenBSD 7.5+, direct `SYS_READV` calls with large `iovec` arrays return `EAGAIN` prematurely. Xray doesn't expect this on a non-empty socket and goes into an infinite wait (hang) without logging any errors.
2. **Implementation (see [HANDOFF.md](https://github.com/user-attachments/files/25847993/HANDOFF.md)):** Once I understood the physics of the bug, I gave the AI strict boundaries to write the workaround. My requirements were: no CGO, create a safe fallback using `syscall.Read` in a loop, preserve your adaptive buffer allocation (1->8), handle `ECONNRESET` correctly to avoid 100% CPU spin loops, and cover 100% of the modified code with tests.

I may not write the code by hand, but I know how to define strict requirements and test hypotheses. This code was statically compiled (CGO=0), deployed on my physical NanoPi Neo3 (aarch64) running OpenBSD 7.8, and it flawlessly resolved the heavy-site hangs in production.

I am attaching my documentation to show that this PR isn't "blind AI generation". 

Thank you for your time and hard work on Xray-core!

**Comment by DavidOsipov** [comment]
Proper fix: https://github.com/golang/go/issues/78049
